From 392a04e4aacd057f39b0b1818c601b94c9e0ea46 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Thu, 4 Sep 2025 16:15:04 +0800
Subject: [PATCH 229/254] mtk: hostapd: add link removal support for multiple
 AP MLDs

The link removal operation used to work only for single AP MLD, as
it utilize the function hostapd_disable_iface() to remove an AP MLD
link.

In setup of multiple AP MLDs, however, this approach affects other
AP MLD links on the same band.

This commit replaces the use of hostapd_disable_iface() with function
hostapd_remove_bss(), along with additional checks for link removal
during such operation.

Also, when generating the ML reconfiguration element, change it to
check whether other links of the same AP MLD have eht_mld_link_removal_count.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/hostapd.c        | 25 ++++++++++++++-----------
 src/ap/ieee802_11_eht.c | 17 +++++++++--------
 2 files changed, 23 insertions(+), 19 deletions(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 986aee784..020f752bd 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -479,7 +479,7 @@ static void hostapd_link_remove_timeout_handler(void *eloop_data,
 
 	hostapd_reset_assoc_stas(hapd);
 	hostapd_free_link_stas(hapd);
-	hostapd_disable_iface(hapd->iface);
+	hostapd_remove_bss(hapd->iface->interfaces, hapd->conf->iface);
 	mld->removed_links &= ~BIT(link_id);
 
 	for_each_mld_link(link, mld->fbss)
@@ -4549,10 +4549,12 @@ int hostapd_remove_bss(struct hapd_interfaces *interfaces, char *buf)
 
 		if (!os_strcmp(hapd_iface->conf->bss[0]->iface, buf)) {
 #ifdef CONFIG_IEEE80211BE
-			if (hostapd_is_mld_ap(hapd_iface->bss[0])) {
-				wpa_printf(MSG_ERROR, "Cannot remove MLD link\n");
-				return -1;
-			}
+			struct hostapd_data *hapd = hapd_iface->bss[0];
+
+			if (hostapd_is_mld_ap(hapd) &&
+			    !(hapd->mld && (hapd->mld->removed_links &
+					    BIT(hapd->mld_link_id))))
+				continue;
 #endif /* CONFIG_IEEE80211BE */
 
 			if (hapd_iface->conf->num_bss == 1) {
@@ -4583,14 +4585,15 @@ int hostapd_remove_bss(struct hapd_interfaces *interfaces, char *buf)
 			}
 		}
 
-		for (j = 0; j < hapd_iface->conf->num_bss; j++) {
+		for (j = 1; j < hapd_iface->conf->num_bss; j++) {
 			if (!os_strcmp(hapd_iface->conf->bss[j]->iface, buf)) {
 #ifdef CONFIG_IEEE80211BE
-				if (hostapd_is_mld_ap(hapd_iface->bss[j])) {
-					wpa_printf(MSG_ERROR,
-						   "Cannot remove MLD link\n");
-					return -1;
-				}
+				struct hostapd_data *hapd = hapd_iface->bss[j];
+
+				if (hostapd_is_mld_ap(hapd) && !(hapd->mld &&
+				    (hapd->mld->removed_links &
+				    BIT(hapd->mld_link_id))))
+					break;
 #endif /* CONFIG_IEEE80211BE */
 
 				hapd_iface->driver_ap_teardown =
diff --git a/src/ap/ieee802_11_eht.c b/src/ap/ieee802_11_eht.c
index 1520bc2ec..60370473f 100644
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -831,22 +831,24 @@ static u8 * hostapd_eid_eht_reconf_ml(struct hostapd_data *hapd, u8 *eid)
 	u16 control;
 	u8 *pos = eid;
 	unsigned int i;
+	bool found = false;
 
 	wpa_printf(MSG_DEBUG, "MLD: Reconfiguration ML");
 
 	/* First check if the element needs to be added */
-	for (i = 0; i < hapd->iface->interfaces->count; i++) {
-		other_hapd = hapd->iface->interfaces->iface[i]->bss[0];
-
-		wpa_printf(MSG_DEBUG, "MLD: Reconfiguration ML: %u",
+	for_each_mld_link(other_hapd, hapd){
+		wpa_printf(MSG_DEBUG, "MLD: link_id %u: Reconfiguration ML: %u",
+			   other_hapd->mld_link_id,
 			   other_hapd->eht_mld_link_removal_count);
 
-		if (other_hapd->eht_mld_link_removal_count)
+		if (other_hapd->eht_mld_link_removal_count) {
+			found = true;
 			break;
+		}
 	}
 
 	/* No link is going to be removed */
-	if (i == hapd->iface->interfaces->count)
+	if (!found)
 		return eid;
 
 	wpa_printf(MSG_DEBUG, "MLD: Reconfiguration ML: Adding element");
@@ -865,8 +867,7 @@ static u8 * hostapd_eid_eht_reconf_ml(struct hostapd_data *hapd, u8 *eid)
 	*pos++ = 1;
 
 	/* Add the per station profiles */
-	for (i = 0; i < hapd->iface->interfaces->count; i++) {
-		other_hapd = hapd->iface->interfaces->iface[i]->bss[0];
+	for_each_mld_link(other_hapd, hapd){
 		if (!other_hapd->eht_mld_link_removal_count)
 			continue;
 
-- 
2.45.2

